src/fetch_document.rs

```diff
@@ -238,182 +42,54 @@ impl Into<Start> for StartPoint {
-        } else {
-            cfg.setup_api().await
-        };
-
-        let data_contract_id = contract_id;
-        tracing::warn!("using existing data contract id and fetching...");
-        let contract_result =
-            DataContract::fetch(&sdk, data_contract_id.clone())
-                .await;
 
-        let contract = match contract_result {
-            Ok(Some(data_contract)) => Arc::new(data_contract),
-            Ok(None) => return Err("data contract not found".to_string()),
-            Err(e) => return Err("fetch data contract error".to_string())
-        };
-
-        tracing::warn!("fetching many...");
-        // Fetch multiple documents so that we get document ID
-        let mut all_docs_query =
-            DocumentQuery::new(Arc::clone(&contract), &document_type)
-                .expect("create SdkDocumentQuery");
-        for wc in where_clauses {
-            all_docs_query = all_docs_query.with_where(wc);
-        }
-        for oc in order_clauses {
-            all_docs_query = all_docs_query.with_order_by(oc);
-        }
-        all_docs_query.limit = limit;
-        all_docs_query.start = match start {
-            Some(s) => Some(s.into()),
-            None => None
-        };
-        tracing::warn!("fetching many... query created");
-        let docs = Document::fetch_many(&sdk, all_docs_query)
-            .await;
-        match docs {
-            Ok(docs) => {
-                tracing::warn!("convert to Vec");
-                let into_vec = |map: BTreeMap<Identifier, Option<Document>>| {
-                    map.into_iter()
-                        .filter_map(|(_key, value)| value)
-                        .collect::<Vec<Document>>()
-                };
-
-                Ok(into_vec(docs))
+fn fetch_documents_with_retry(
+    sdk: Arc<Sdk>,  // No need for a reference; pass the Arc by value
+    data_contract_cache: Arc<Cache<Identifier, DataContract>>,
+    query: DocumentQuery,
+    request_settings: RequestSettings,
+    retries_left: usize,
+) -> BoxFuture<'static, Result<Documents, Error>> {
+    // Clone the Arc<Sdk> here to ensure it's owned by the future
+    Box::pin(async move {
+        match Document::fetch_many_with_settings(&sdk, query.clone(), request_settings).await {
+            Ok(documents) => Ok(documents),
+            Err(error) => {
+                if retries_left > 1 {
+                    if error.to_string().contains("contract not found error: contract not found when querying from value with contract info") {
+                        if (data_contract_cache.get(&query.data_contract.id()) != None) {
+                            return fetch_documents_with_retry(sdk, data_contract_cache, query, request_settings, retries_left - 1).await;
+                        }
+                    }
+                }
+                Err(error)
             }
-            Err(e) => Err(e.to_string())
         }
     })
 }
```

Comment:
This will look for a "contract not found" error and recursively call itself, hopefully that will choose another EvoNode to make the DAPI call to.